docs/DafnyRef/UserGuide.md

```diff
@@ -1147,25 +1147,30 @@ method innerLoop()
 
 In the next section, when everything can be proven in a timely manner, we explain another strategy to decrease proof time by parallelizing it if needed, and making the verifier focus on certain parts.
 
-### 13.6.3. Assertion batches {#sec-assertion-batches}
+### 13.6.3. Assertion batches, well-formedness, correctness {#sec-assertion-batches}
 
 To understand how to control verification,
 it is first useful to understand how `dafny` verifies functions and methods.
 
-For every method (or function, constructor, etc.), `dafny` extracts _assertions_. Here is a non-exhaustive list of such extracted assertions:
+For every method (or function, constructor, etc.), `dafny` extracts _assertions_.
+Assertions can either be *implicit* or *explicit*.
+Implicit assertions, also called *well-formedness assertions*, ensure expressions make sense.
+Explicit assertions, also called *correctness assertions*, are provided by the user, and can either help prove implicit assertions, or prove other desired higher-level properties.
```

Comment:
```suggestion
Explicit assertions, also called *correctness assertions*, are provided by the user, and can either help prove other assertions, or the specification.
```
Do we use the term _correctness assertion_ elsewhere?

Reply:
my new version:
Explicit assertions, also called *correctness assertions*, are provided by the user, and can help prove other implicit and explicit assertions.

> Do we use the term correctness assertion elsewhere?

yes in boogie 😸

Reply:
We have the duality implicit/explicit assertion.
We have a term to qualify checking implicit assertions, which is "well-formedness check". But often that confuses users because they don't make the distinction, for them Dafny is just verifying everything.
So we need another term to say that an assertion is not part of the well-formedness check, and it's the correctness check, which is the official term we use when turning on logging in the CLI.

Reply:
As well-formedness does not convey much meaning to you, I recommend you read the following section of the documentation which use the term "well-formed" in context with examples and might give you some intuition about why this term exist.
https://dafny.org/dafny/DafnyRef/DafnyRef#sec-implication-and-reverse-implication
https://dafny.org/dafny/DafnyRef/DafnyRef#sec-multi-dimensional-arrays
https://dafny.org/dafny/DafnyRef/DafnyRef#sec-inductive-datatypes

Under the hood, this term is different from regular correctness because expressions in Dafny are translated almost raw into mathematical formulas, and formulas can be constructed only if we can independently prove they make sense in the first place, and this step is called well-formedness check.

> Also, what's the reason we need a separate correctness and well-formedness check again? 

It should be totally doable and equivalent to do the verification of both things in a single implementation, and we probably should down the road as there are ways to do that in Boogie. But merging the two checks in a single Boogie implementation does not intrinsically make them the same thing.

Moreover, the distinction between well-formedness and correctness is a bit blurred. For example, if there is an explicit assert StmtExpr as part of the requires or ensures clause, it is still checked during the well-formedness of the pre/post condition even though it's explicit. For expressions, this distinction exists as well, although well-formedness is checked in the same scope as correctness.

Reply:
> I recommend you read the following section of the documentation which use the term "well-formed" in context with examples and might give you some intuition about why this term exist.

Thanks

Do I understand correctly that code is well-formed if it satisfies all requires clauses of the calls it makes, be it calls to native code such as indexing and division, or calls to user-code? If the answer is yes, then we should add that definition to the documentation.

> For example, if there is an explicit assert StmtExpr as part of the requires or ensures clause, it is still checked during the well-formedness of the pre/post condition even though it's explicit. For expressions, this distinction exists as well, although well-formedness is checked in the same scope as correctness.

I think that's to be expected, because a requires clause translates to an assume statement in the callee, and all calls in the body of the callee are part of its WF check, including that assume statement.

Reply:
> Do I understand correctly that code is well-formed if it satisfies all requires clauses of the calls it makes, be it calls to native code such as indexing and division, or calls to user-code? 

That made me realize that well-formedness makes a distinction between expressions and statements, which are both code. There is no well-formedness of statements (this is correctness) and there is no correctness of expressions (this is well-formedness).
I'll reword it

Reply:
What do you mean with there is no well-formedness of statements? If I make a call to a method, which is a statement, then that introduces a check in the well-formedness Boogie `implementation`, does it not?

Reply:
What check does it introduce in the case you mention?
Note that I'm taking expressions in a large sense, for example ExprRHS is an expression, so for a method call to be validated, one has to prove its well-formedness, i.e. the requires clause holds. But that's not the well-formedness of the statement, only expressions.
Similarly, when having an assignment like A.x := 3;, A is an expression that computes an object, and a part of the well-formedness is that A is not null.

Reply:
Okay.. well, regardless of what we call expressions or statements, how do feel about what I said before?

> Do I understand correctly that code is well-formed if it satisfies all requires clauses of the calls it makes, be it calls to native code such as indexing and division, or calls to user-code? 

> If the answer is yes, then should we add that definition to the documentation? It seems the documentation currently only defines what well-formedness is by example. Also, I would consider replacing the term "well-formed" with "satisfies requires clauses" which is more verbose but self-explanatory.

Reply:
Happy to make the documentation more user-meaningful with your help.

> Do I understand correctly that [expressions are] well-formed if [they satisfy] all requires clauses of the calls it makes, be it calls to native code such as indexing and division, or calls to user-code?

I would say yes, that's about that. That way, it's possible to define the expression as a mathematical expression without having to bother that it assumes something that is wrong.

Looking at the documentation, when it says `a.Length < 0 ==> a == null  // not well-formed`, it becomes evident that the formula to not be-well-formed requires that that there is no assumption before that, as if it was the first requires clause on a function, which we don't know out of context. Hence, I realize that the example `a.Length < 0 ==> a == null  // not well-formed` is not well-formed :-) and well-formedness also depends on the context.

> If the answer is yes, then should we add that definition to the documentation? It seems the documentation currently only defines what well-formedness is by example. Also, I would consider replacing the term "well-formed" with "satisfies requires clauses" which is more verbose but self-explanatory.

Let me work on that.